push-image.sh

```diff
@@ -1,76 +1,94 @@
 #!/usr/bin/env bash
-set -e
+set -euo pipefail
 
-# Push the 'conjur' image to various Docker registries
-# Push tagged images on master branch
-# Release images can be created by passing the desired tag to this script
-# Ex: ./push-image 4.9.5.1
+# Pushes the 'conjur' image to various Docker registries
+# There are essentially three flows this script supports
+# - Pushing images to the internal registry during the early pipeline stage
+# - Pushing an "edge"-tagged image to DockerHub
+# - Publishing images to public registries on new project releases during a
+#   tag-triggered build
+#
+# The default flow when no parameters are supplied is to publish images to
+# public registries. This does require that the TAG_NAME variable is set, which
+# happens automatically in tag-triggered builds.
+#
+# To publish to a (private) internal registry, run the script with the "-r" flag
+# to supply the registry prefix.
+#
+# To publish an "edge"-tagged release to DockerHub, run the script with the "-e"
+# flag.
 
 # shellcheck disable=SC1091
-. version_utils.sh
+. build_utils.sh
 
-if [[ -z "${TAG_NAME:-}" ]]; then
-  echo "Please supply environment variable TAG_NAME."
-  echo "If you see this error in Jenkins it means the publish script was run"
-  echo "for a build that wasn't triggered by a tag -" \
-    "please check publish stage conditions."
-  exit 1
-fi
-
-TAG="${1:-$(version_tag)}"
-VERSION="$(< VERSION)"
-SOURCE_IMAGE="conjur:$TAG"
-RH_SOURCE_IMAGE="conjur-ubi:$TAG"
+function print_help() {
+  echo "Usage: $0 [ -r <registry prefix> ] [ -e ]"
+  echo " -r optional registry prefix to prepend to '/cyberark/conjur'"
+  echo " -e optional flag to publish edge-tagged image to DockerHub"
+}
 
-CONJUR_REGISTRY="registry.tld"
+REGISTRY_PREFIX=""
+PUBLISH_EDGE=false
+TAG="$(version_tag)"
+LOCAL_IMAGE="conjur:${TAG}"
+RH_LOCAL_IMAGE="conjur-ubi:${TAG}"
 IMAGE_NAME="cyberark/conjur"
 REDHAT_IMAGE="scan.connect.redhat.com/ospid-9fb7aea1-0c01-4527-8def-242f3cde7dc6/conjur"
 
-# both old-style 'conjur' and new-style 'cyberark/conjur'
-INTERNAL_IMAGES=("$CONJUR_REGISTRY/conjur" "$CONJUR_REGISTRY/$IMAGE_NAME")
+while getopts "r:e" opt; do
+    case "${opt}" in
+        r )
+          REGISTRY_PREFIX="${OPTARG}"
+          ;;
+        e )
+          PUBLISH_EDGE=true
+          ;;
+       \? )
```

Comment:
Where does the `\?` come from? Does this mean you would use `push-image.sh \?` to display the help text, and if so, what is the value of this over `--help|-h`?

Does it need to be in the `getopts` above as well?

Reply:
I find [this blog](https://sookocheff.com/post/bash/parsing-bash-script-arguments-with-shopts/) to be a helpful reference when I'm using getopts. in it the author says:
> if an invalid option is provided, the option variable is assigned the value ?. You can catch this case and provide an appropriate usage message to the user.

I could also add a help flag, but I'm not sure it's worth it since I doubt this will be run manually very much.

Reply:
Ah, got it. And thanks for sharing the article on that, TIL. I agree a separate help flag is unnecessary. 

Thanks!